# <center>Guided Project: Exploring eBay Car Sales Data</center>
## <center>Monique C. Wilson</center>

This is a guided project from the Dataquest.io Data Analyst in Python certification course.  The project specifically covers exploring data using pandas and numpy.

The dataset used was scrapped and uploaded to Kaggle.  It contains info about used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website. 

For the purposes of this project, a modified ("dirtied") and shortened version of the dataset is used.

The data dictionary provided with data is as follows:
<ul>
    <li>dateCrawled - When this ad was first crawled. All field-values are taken from this date.</li>
<li>name - Name of the car.</li>
<li>seller - Whether the seller is private or a dealer.</li>
<li>offerType - The type of listing</li>
<li>price - The price on the ad to sell the car.</li>
<li>abtest - Whether the listing is included in an A/B test.</li>
<li>vehicleType - The vehicle Type.</li>
<li>yearOfRegistration - The year in which which year the car was first registered.</li>
<li>gearbox - The transmission type.</li>
<li>powerPS - The power of the car in PS.</li>
<li>model - The car model name.</li>
<li>kilometer - How many kilometers the car has driven.</li>
<li>monthOfRegistration - The month in which which year the car was first registered.</li>
<li>fuelType - What type of fuel the car uses.</li>
<li>brand - The brand of the car.</li>
<li>notRepairedDamage - If the car has a damage which is not yet repaired.</li>
<li>dateCreated - The date on which the eBay listing was created.</li>
<li>nrOfPictures - The number of pictures in the ad.</li>
<li>postalCode - The postal code for the location of the vehicle.</li>
<li>lastSeenOnline - When the crawler saw this ad last online.</li>
</ul>
The aim of this project is to clean the data and analyze the included used car listings.

In [1]:
import pandas as pd
import numpy as np

autos = pd.read_csv("autos.csv", encoding="latin1")

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [3]:
autos.info

<bound method DataFrame.info of                dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V

In [4]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [5]:
autos["nrOfPictures"].value_counts()

nrOfPictures
0    50000
Name: count, dtype: int64

The first thing to note about this dataset is that it is in German.  There appear to be data inconsistencies in multiple columns.  For example, the vehicleType, model, and notRepairedDamage columns should be strings, yet they have NaN as some of the values.  The monthOfRegistration column consists of integers, however some entries have a value of 0.  Despite this being a tool to sell used cars, it appears none of the entries are accompanied by photos.

# <center>Cleaning the Columns</center>

The preferred convention for Python strings is snakecase.  However, the column names in this dataset use camelcase. 

In [6]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

These need to be changed. Some simply need to go from camelcase to snakecase. Others need to be reworded for clarity.

In [7]:
new_columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_pictures', 'postal_code',
       'last_seen']
autos.columns = new_columns
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

# <center>Initial Exploration and Cleaning of Data</center>

The initial exploration of the data will help to inform on the cleaning tasks that will be required for the data.  Some things to look for include, but are not limited to:

<ul>
    <li>Columns containing the same value for most or all entries can be dropped as they are not useful for analysis</li>
    <li>Numeric data stored as text needs to be cleaned and converted</li>
</ul>

This process was started earlier when using Series.value_counts() to determine that the num_pictures (previously nrOfPictures) contains the value of 0 for all the entries.  As such, that column will be dropped.

In [8]:
autos.describe(include="all")

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_pictures   postal_code            last_seen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

In addition to the num_pictures column, the seller and offer_type columns can also be dropped.

In [9]:
autos = autos.drop(["seller", "num_pictures", "offer_type"], axis=1)
autos.columns

Index(['date_crawled', 'name', 'price', 'abtest', 'vehicle_type',
       'registration_year', 'gearbox', 'power_ps', 'model', 'odometer',
       'registration_month', 'fuel_type', 'brand', 'unrepaired_damage',
       'ad_created', 'postal_code', 'last_seen'],
      dtype='object')

Several columns that should have only numeric values seem to be represented as strings or include non-numeric characters.  These will need to be examined further.
<ul>
    <li>price</li>
    <li>registration_year</li>
    <li>power_ps</li>
    <li>odometer</li>
    <li>registration_month</li>
</ul>


### Price

In [10]:
autos["price"].unique()

array(['$5,000', '$8,500', '$8,990', ..., '$385', '$22,200', '$16,995'],
      dtype=object)

The $ symbol needs to be stripped so that only a numeric value remains. It looks like the commas also need to be removed from the price column. The dtype is currently object and needs to be a numeric type, such as int or float.

In [11]:
autos["price"] = (autos["price"]
                          .str.replace("$","")
                          .str.replace(",","")
                          .astype(int)
                          )
autos["price"].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int32

Usually, a price column would be of type float because monetary values include decimal points.  In this case, all of the dollar values appear to have been rounded so the integer type was sufficient.

The same exploration process can be followed for the remaining columns in question.

### Registration Year

In [12]:
autos["registration_year"].unique()

array([2004, 1997, 2009, 2007, 2003, 2006, 1995, 1998, 2000, 2017, 2010,
       1999, 1982, 1990, 2015, 2014, 1996, 1992, 2005, 2002, 2012, 2011,
       2008, 1985, 2016, 1994, 1986, 2001, 2018, 2013, 1972, 1993, 1988,
       1989, 1967, 1973, 1956, 1976, 4500, 1987, 1991, 1983, 1960, 1969,
       1950, 1978, 1980, 1984, 1963, 1977, 1961, 1968, 1934, 1965, 1971,
       1966, 1979, 1981, 1970, 1974, 1910, 1975, 5000, 4100, 2019, 1959,
       9996, 9999, 6200, 1964, 1958, 1800, 1948, 1931, 1943, 9000, 1941,
       1962, 1927, 1937, 1929, 1000, 1957, 1952, 1111, 1955, 1939, 8888,
       1954, 1938, 2800, 5911, 1500, 1953, 1951, 4800, 1001], dtype=int64)

It looks like all of the values in the registration_year column are actually integers.  However, there are several values that do not make sense for a vehicle registration year: 4500, 5000, 4100, 9996, 9999, 6200, 1800, 9000, 1000, 1111, 8888, 2800, 5911, 1500, 4800, and 1001.  That is 16 values out of 50,000 values. These will be dealt with later. Now it's time to move to the next column in need of exploration.


### Odometer

In [13]:
autos["odometer"].unique()

array(['150,000km', '70,000km', '50,000km', '80,000km', '10,000km',
       '30,000km', '125,000km', '90,000km', '20,000km', '60,000km',
       '5,000km', '100,000km', '40,000km'], dtype=object)

Similar to the price column, the data here needs to be stripped of commas and letters.  It then needs to be converted to a numeric dtype instead of an object dtype.  The name of the column will need to be renamed to include the "km" unit of measurement.

In [14]:
autos["odometer"] = (autos["odometer"]
                          .str.replace("km","")
                          .str.replace(",","")
                          .astype(int)
                          )
autos.rename({"odometer": "odometer_km"}, axis=1, inplace=True)
autos["odometer_km"].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int32

Now the odometer column contains uniform, relevant data and is ready for further exploration.  For now, it's time to move to the next column.

### Registration Month

In [15]:
autos["registration_month"].unique()

array([ 3,  6,  7,  4,  8, 12, 10,  0,  9, 11,  5,  2,  1], dtype=int64)

In [16]:
autos["registration_month"].value_counts()

registration_month
0     5075
3     5071
6     4368
5     4107
4     4102
7     3949
10    3651
12    3447
9     3389
11    3360
1     3282
8     3191
2     3008
Name: count, dtype: int64

There are 12 months in a year. So it only makes sense to have the values 1 - 12 in this column.  Again, these invalid values will be dealt with later upon further exploration.


# <center>Further Exploration and Cleaning</center>

Now it's time to make some decisions about the data issues that have been noted.

### Price

Further exploration will determine if there are any pricing anomalies or outliers.

In [17]:
print(autos["price"].unique().shape)
print(autos["price"].describe())
autos["price"].value_counts().head(20)

(2357,)
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


price
0       1421
500      781
1500     734
2500     643
1000     639
1200     639
600      531
800      498
3500     498
2000     460
999      434
750      433
900      420
650      419
850      410
700      395
4500     394
300      384
2200     382
950      379
Name: count, dtype: int64

In [18]:
autos["price"].value_counts().sort_index().head(20)

price
0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: count, dtype: int64

In [19]:
autos["price"].value_counts().sort_index(ascending=False).head(20)

price
99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: count, dtype: int64

There are 1,421 cars priced at \\$0 and one at \\$99,999,999. The cars are listed on eBay, which is known for selling items via auctions.  It would make sense that an auction would have a starting bid of \\$1, but not \\$0 or \\$99,999,999.  Furthermore, there is no realistic situation in which a car would be sold for \\$99,999,999.  In fact, the most expensive car on record was sold for \\$142 million.  It was a 1955 Mercedes-Benz 300 SLR Uhlenhaut Coupe sold at auction.  Many ultra luxury cars go for around \\$40 million.  Any values below \\$1 or above \\$200,000 can be excluded from the dataset.

In [20]:
autos = autos[autos["price"].between(1,201000)]
autos["price"].describe()

count     48557.000000
mean       5842.888914
std        8301.871663
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7450.000000
max      198000.000000
Name: price, dtype: float64

Now, the remaining data only includes cars that have a starting price between \\$1 and \\$200,000.

### Registration Year

German history includes three prominent dates with regards to car registrations:

    -1896: introduction a a vehicular registration system to Germany
    -1906: registered plates mandated
    -1956: current plate registration system introduced

With that in mind, any listing with a registration date prior to 1900 can probably be removed.  This data was collected in 2016. Any listing with a date after that can not possibly be valid and can also probably be removed. First, it needs to be determined how much of an impact the removal of this data would have on the dataset.

In [21]:
(~autos["registration_year"].between(1900,2016)).sum() / autos.shape[0]

0.03879976110550487

The data in question represents 3.87% of the overall dataset. Removal should not impact the data.

In [22]:
autos = autos[autos["registration_year"].between(1900,2016)]
autos["registration_year"].value_counts(normalize=True).head(10)

registration_year
2000    0.067619
2005    0.062906
1999    0.062070
2004    0.057914
2003    0.057828
2006    0.057207
2001    0.056457
2002    0.053264
1998    0.050629
2007    0.048786
Name: proportion, dtype: float64

Before moving on to other columns, it would be good to review what data remains to see if any other problems have been remedied.

In [26]:
autos["registration_month"].unique()

array([ 3,  6,  7,  4,  8, 12, 10,  0,  9, 11,  5,  2,  1], dtype=int64)

In [27]:
autos["registration_month"].value_counts()

registration_month
3     4836
6     4117
0     4025
4     3894
5     3877
7     3720
10    3495
12    3262
9     3243
11    3229
1     3105
8     3020
2     2850
Name: count, dtype: int64

In [29]:
(~autos["registration_month"].between(1,12)).sum() / autos.shape[0]

0.08623829623122577

There are still 4,025 listings with a registration month of 0, which is invalid. However, that data represents almost 10% of the dataset.  It would not be wise to remove it. Instead, any listing with the value of 0 for the registration month will be changed to the last month of the year, 12.

In [31]:
mapping_dict = {
    0:12,
    1:1,
    2:2,
    3:3,
    4:4,
    5:5,
    6:6,
    7:7,
    8:8,
    9:9,
    10:10,
    11:11,
    12:12
    }
autos["registration_month"] = autos["registration_month"].map(mapping_dict)

In [32]:
autos["registration_month"].unique()

array([ 3,  6,  7,  4,  8, 12, 10,  9, 11,  5,  2,  1], dtype=int64)

Now the date data can be reviewed for observation.

### Exploration of Other Date Columns

In [33]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [34]:
(autos["date_crawled"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

date_crawled
2016-03-05    0.025197
2016-03-06    0.014162
2016-03-07    0.036252
2016-03-08    0.033553
2016-03-09    0.033253
2016-03-10    0.032246
2016-03-11    0.032460
2016-03-12    0.036831
2016-03-13    0.015876
2016-03-14    0.036338
2016-03-15    0.034367
2016-03-16    0.029503
2016-03-17    0.031796
2016-03-18    0.012813
2016-03-19    0.034667
2016-03-20    0.038031
2016-03-21    0.037302
2016-03-22    0.032846
2016-03-23    0.032181
2016-03-24    0.029482
2016-03-25    0.031517
2016-03-26    0.032074
2016-03-27    0.030767
2016-03-28    0.034560
2016-03-29    0.034110
2016-03-30    0.033788
2016-03-31    0.031796
2016-04-01    0.033810
2016-04-02    0.035567
2016-04-03    0.038759
2016-04-04    0.036616
2016-04-05    0.012984
2016-04-06    0.003085
2016-04-07    0.001414
Name: proportion, dtype: float64

In [35]:
(autos["date_crawled"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_values()
        )

date_crawled
2016-04-07    0.001414
2016-04-06    0.003085
2016-03-18    0.012813
2016-04-05    0.012984
2016-03-06    0.014162
2016-03-13    0.015876
2016-03-05    0.025197
2016-03-24    0.029482
2016-03-16    0.029503
2016-03-27    0.030767
2016-03-25    0.031517
2016-03-31    0.031796
2016-03-17    0.031796
2016-03-26    0.032074
2016-03-23    0.032181
2016-03-10    0.032246
2016-03-11    0.032460
2016-03-22    0.032846
2016-03-09    0.033253
2016-03-08    0.033553
2016-03-30    0.033788
2016-04-01    0.033810
2016-03-29    0.034110
2016-03-15    0.034367
2016-03-28    0.034560
2016-03-19    0.034667
2016-04-02    0.035567
2016-03-07    0.036252
2016-03-14    0.036338
2016-04-04    0.036616
2016-03-12    0.036831
2016-03-21    0.037302
2016-03-20    0.038031
2016-04-03    0.038759
Name: proportion, dtype: float64

In [36]:
(autos["last_seen"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

last_seen
2016-03-05    0.001071
2016-03-06    0.004114
2016-03-07    0.005378
2016-03-08    0.007478
2016-03-09    0.009770
2016-03-10    0.010691
2016-03-11    0.012384
2016-03-12    0.023761
2016-03-13    0.008656
2016-03-14    0.012663
2016-03-15    0.016005
2016-03-16    0.016284
2016-03-17    0.028089
2016-03-18    0.007220
2016-03-19    0.015619
2016-03-20    0.020633
2016-03-21    0.020590
2016-03-22    0.020847
2016-03-23    0.018362
2016-03-24    0.019690
2016-03-25    0.018940
2016-03-26    0.016798
2016-03-27    0.015619
2016-03-28    0.020676
2016-03-29    0.022090
2016-03-30    0.024618
2016-03-31    0.023633
2016-04-01    0.022947
2016-04-02    0.024661
2016-04-03    0.025132
2016-04-04    0.024125
2016-04-05    0.125383
2016-04-06    0.223320
2016-04-07    0.132753
Name: proportion, dtype: float64

The last_seen column data can provide an indication of when the car was sold, with the exception of the last few days.  The drastic increase in "last seen" posts may be more related to the end of the ad campaign and not due to car sales.

# <center>Exploring Price By Brand</center>

As the heading suggests, the data can be aggregated to explore the variations in pricing across different car brands.

In [37]:
autos["brand"].value_counts(normalize=True)

brand
volkswagen        0.211300
bmw               0.110021
opel              0.107600
mercedes_benz     0.096480
audi              0.086581
ford              0.069912
renault           0.047158
peugeot           0.029846
fiat              0.025647
seat              0.018276
skoda             0.016412
nissan            0.015276
mazda             0.015191
smart             0.014162
citroen           0.014012
toyota            0.012705
hyundai           0.010027
sonstige_autos    0.009792
volvo             0.009149
mini              0.008763
mitsubishi        0.008227
honda             0.007842
kia               0.007070
alfa_romeo        0.006642
porsche           0.006021
suzuki            0.005935
chevrolet         0.005699
chrysler          0.003514
dacia             0.002635
daihatsu          0.002507
jeep              0.002271
subaru            0.002143
land_rover        0.002100
saab              0.001650
jaguar            0.001564
daewoo            0.001500
trabant           0.00

Almost half of the brands represented are German brands, with Volkswagen being the top brand.  While there are 40 brands represented in the dataset, the focus will be on those brands that represent more than 4% of total listings.

In [38]:
brand_counts = autos["brand"].value_counts(normalize=True)
brands_used = brand_counts[brand_counts > .04].index
print(brands_used)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford',
       'renault'],
      dtype='object', name='brand')


In [39]:
price_by_brand_means = {}

for brand in brands_used:
    brand_only = autos[autos["brand"] == brand]
    mean_price = brand_only["price"].mean()
    price_by_brand_means[brand] = int(mean_price)

price_by_brand_means

{'volkswagen': 5402,
 'bmw': 8236,
 'opel': 2975,
 'mercedes_benz': 8628,
 'audi': 9336,
 'ford': 3749,
 'renault': 2474}

Volkswagen, the most popular brand represented in the dataset, seems to be in the "Goldilocks" price range.  It's not as expensive as higher-end brands, such as Mercedes-Benz and Audi.  It's also not as cheap as the lower-end brands, such as Renault and Opel.

## <center>Exploring Average Mileage</center>

Data aggregation techniques can also be used to analyze mileage with respect to the brands explored above.

In [41]:
bmp_series = pd.Series(price_by_brand_means)
print(bmp_series)

volkswagen       5402
bmw              8236
opel             2975
mercedes_benz    8628
audi             9336
ford             3749
renault          2474
dtype: int64


In [42]:
# use the newly created series to create a one-column dataframe
df = pd.DataFrame(bmp_series, columns=['mean_price'])
df

mean_price
volkswagen           5402
bmw                  8236
opel                 2975
mercedes_benz        8628
audi                 9336
ford                 3749
renault              2474

In [43]:
mileage_by_brand_mean = {}

for brand in brands_used:
    brand_only = autos[autos["brand"] == brand]
    mean_mileage = brand_only["odometer_km"].mean()
    mileage_by_brand_mean[brand] = int(mean_mileage)

mean_mileage = pd.Series(mileage_by_brand_mean).sort_values(ascending=False)
mean_prices = pd.Series(price_by_brand_means).sort_values(ascending=False)
brand_info = pd.DataFrame(mean_mileage,columns=['mean_mileage'])
brand_info

mean_mileage
bmw                  132619
mercedes_benz        130788
opel                 129310
audi                 129157
volkswagen           128707
renault              128071
ford                 124266

In [44]:
# add mean_price column

brand_info["mean_price"] = mean_prices
brand_info

mean_mileage  mean_price
bmw                  132619        8236
mercedes_benz        130788        8628
opel                 129310        2975
audi                 129157        9336
volkswagen           128707        5402
renault              128071        2474
ford                 124266        3749

Though it seems counterintuitive, it appears as though the vehicles costing more had higher mileage than the vehicles that were less expensive.